test/e2e/kubectl/kubectl.go

```diff
@@ -1079,6 +1106,111 @@ metadata:
+							OpenAPIV3Schema: props,
+						},
+					},
+				}
+			}
+
+			group := fmt.Sprintf("%s.example.com", f.BaseName)
+			testCRD, err := crd.CreateMultiVersionTestCRD(f, group, opt)
+			if err != nil {
+				framework.Failf("failed to create test CRD: %v", err)
+			}
+			defer testCRD.CleanUp()
+
+			ginkgo.By("sleep for 10s to wait for potential crd openapi publishing alpha feature")
+			time.Sleep(10 * time.Second)
+
+			ginkgo.By("attempting to create a CR with unknown metadata fields at the root level")
+			gvk := schema.GroupVersionKind{Group: testCRD.Crd.Spec.Group, Version: testCRD.Crd.Spec.Versions[0].Name, Kind: testCRD.Crd.Spec.Names.Kind}
+			schema := schemaForGVK(gvk)
+			framework.ExpectNotEqual(schema, nil, "retrieving a schema for the crd")
+			embeddedCRPattern := `
+
+{%s,
+  "spec": {
+    "template": {
+      "apiVersion": "foo/v1",
+      "kind": "Sub",
+      "metadata": {
+        %s
+        "name": "subobject",
+        "namespace": "my-ns"
+      }
+    }
+  }
+}`
+			meta := unknownFieldMetadataJSON(gvk, "test-cr")
+			unknownRootMetaCR := fmt.Sprintf(embeddedCRPattern, meta, "")
+			_, err = framework.RunKubectlInput(ns, unknownRootMetaCR, "create", "--validate=true", "-f", "-")
+			if !(strings.Contains(err.Error(), `unknown field "unknownMeta"`) || strings.Contains(err.Error(), `unknown field "metadata.unknownMeta"`)) {
+				framework.Failf("error missing root unknown metadata field, got: %v", err)
+			}
+
+			ginkgo.By("attempting to create a CR with unknown metadata fields in the embedded object")
+			metaEmbedded := fmt.Sprintf(metaPattern, testCRD.Crd.Spec.Names.Kind, testCRD.Crd.Spec.Group, testCRD.Crd.Spec.Versions[0].Name, "test-cr-embedded")
+			unknownEmbeddedMetaCR := fmt.Sprintf(embeddedCRPattern, metaEmbedded, `"unknownMetaEmbedded": "bar",`)
+			_, err = framework.RunKubectlInput(ns, unknownEmbeddedMetaCR, "create", "--validate=true", "-f", "-")
+			if !(strings.Contains(err.Error(), `unknown field "unknownMetaEmbedded"`) || strings.Contains(err.Error(), `unknown field "spec.template.metadata.unknownMetaEmbedded"`)) {
+				framework.Failf("error missing embedded unknown metadata field, got: %v", err)
+			}
+		})
+
+		ginkgo.It("should detect unknown metadata fields of a typed object", func() {
+			ginkgo.By("calling kubectl create deployment")
+			invalidMetaDeployment := `
+	{
+		"apiVersion": "apps/v1",
+		"kind": "Deployment",
+		"metadata": {
+			"name": "my-dep",
+			"namespace": "my-ns",
+			"unknownMeta": "foo",
+			"labels": {"app": "nginx"}
+		},
+		"spec": {
+			"selector": {
+				"matchLabels": {
+					"app": "nginx"
+				}
+			},
+			"template": {
+				"metadata": {
+					"labels": {
+						"app": "nginx"
+					}
+				},
+				"spec": {
+					"containers": [{
+						"name":  "nginx",
+						"image": "nginx:latest"
+					}]
+				}
+			}
+		}
+	}
+		`
+			_, err := framework.RunKubectlInput(ns, invalidMetaDeployment, "create", "-f", "-")
```

Comment:
before calling err.Error(), assert err is non-nil (to avoid an unclear panic error)